linkerd/drain/src/lib.rs

```diff
@@ -1,95 +1,97 @@
 #![deny(warnings, rust_2018_idioms)]
 
+#[cfg(feature = "tower")]
 mod retain;
 
+#[cfg(feature = "tower")]
 pub use crate::retain::Retain;
-use linkerd2_error::Never;
-use pin_project::pin_project;
-use std::{
-    future::Future,
-    pin::Pin,
-    task::{Context, Poll},
-};
-use tokio::{
-    stream::Stream,
-    sync::{mpsc, watch},
-};
+use std::future::Future;
+use tokio::sync::{mpsc, watch};
 
 /// Creates a drain channel.
 ///
 /// The `Signal` is used to start a drain, and the `Watch` will be notified
 /// when a drain is signaled.
 pub fn channel() -> (Signal, Watch) {
-    let (tx, rx) = watch::channel(());
+    let (signal_tx, signal_rx) = watch::channel(());
     let (drained_tx, drained_rx) = mpsc::channel(1);
 
-    (Signal { drained_rx, tx }, Watch { drained_tx, rx })
+    let signal = Signal {
+        drained_rx,
+        signal_tx,
+    };
+    let watch = Watch {
+        drained_tx,
+        signal_rx,
+    };
+    (signal, watch)
 }
 
+enum Never {}
+
 /// Send a drain command to all watchers.
-///
-/// When a drain is started, this returns a `Drained` future which resolves
-/// when all `Watch`ers have been dropped.
 #[derive(Debug)]
 pub struct Signal {
     drained_rx: mpsc::Receiver<Never>,
-    tx: watch::Sender<()>,
+    signal_tx: watch::Sender<()>,
 }
 
 /// Watch for a drain command.
 ///
-/// This wraps another future and callback to be called when drain is triggered.
-#[pin_project]
+/// All `Watch` instances must be dropped for a `Signal::signal` call to
+/// complete.
 #[derive(Clone, Debug)]
 pub struct Watch {
-    #[pin]
     drained_tx: mpsc::Sender<Never>,
-    rx: watch::Receiver<()>,
-}
-
-/// A future that resolves when all `Watch`ers have been dropped (drained).
-#[pin_project]
-pub struct Drained {
-    #[pin]
-    drained_rx: mpsc::Receiver<Never>,
+    signal_rx: watch::Receiver<()>,
 }
 
 #[must_use = "ReleaseShutdown should be dropped explicitly to release the runtime"]
 #[derive(Clone, Debug)]
 pub struct ReleaseShutdown(mpsc::Sender<Never>);
 
-// ===== impl Signal =====
+// === impl Signal ===
 
 impl Signal {
     /// Start the draining process.
     ///
     /// A signal is sent to all futures watching for the signal. A new future
     /// is returned from this method that resolves when all watchers have
     /// completed.
```

Comment:
This comment is no longer exactly correct; the behavior here changed very slightly. previously, the drain signal was sent _immediately_ when calling the function, and then a future that waits for the drain to complete is returned. Now that this is an `async fn`, that behavior is slightly different: the drain isn't signalled until the future is awaited. We should either update the comment, or change the method to preserve the old behavior. 

If we wanted to keep the old behavior, we could still use async/await syntax, like so:
```rust
pub fn drain(mut self) -> impl Future<Output = ()> + Send + Sync {        
    // Update the state of the signal watch so that all watchers are observe
    let _ = self.signal_tx.send(());

    async move {
        // Wait for all watchers to release their drain handle.
        match self.drained_rx.recv().await {
            None => {}
            Some(n) => match n {},
        }
    }
}
```
This still signals the drain *before* returning a future that can be awaited.

Alternatively, unless this behavior is relied on (and I don't *think* it is) we can just change the comment. :)